modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/AsyncWrapper.java

```diff
@@ -0,0 +1,149 @@
+
+    private volatile boolean cancelled = false;
+
+    private volatile boolean firstRequest = true;
+
+    /**
+     * Constructor.
+     *
+     * <p>The execution will be in the thread invoking particular method of this cursor.
+     *
+     * @param source An iterator to wrap.
+     */
+    public AsyncWrapper(Iterator<T> source) {
+        this(CompletableFuture.completedFuture(source), Runnable::run);
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param initFut Initialization future.
+     * @param exec An executor to delegate execution.
+     */
+    public AsyncWrapper(CompletableFuture<Iterator<T>> initFut, Executor exec) {
+        this.cursorFut = initFut;
+        this.exec = exec;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletionStage<BatchedResult<T>> requestNext(int rows) {
+        CompletableFuture<BatchedResult<T>> next = new CompletableFuture<>();
+        CompletableFuture<BatchedResult<T>> prev;
+
+        synchronized (lock) {
+            if (cancelled) {
+                next.completeExceptionally(new ClosedCursorException());
+
+                return next;
+            }
+
+            prev = requestChainTail;
+            requestChainTail = next;
+        }
+
+        prev.thenCompose(tmp -> cursorFut).thenAcceptAsync(cursor -> {
+            int remains = rows;
+
+            if (!cursor.hasNext() && !firstRequest) {
+                next.completeExceptionally(new NoSuchElementException());
+
+                return;
+            }
+
+            List<T> batch = new ArrayList<>(rows);
+
+            firstRequest = false;
+
+            while (remains-- > 0 && cursor.hasNext()) {
+                batch.add(cursor.next());
+            }
+
+            next.complete(new BatchedResult<>(batch, cursor.hasNext()));
+        }).exceptionally(t -> {
+            next.completeExceptionally(t);
+
+            return null;
+        });
+
+        return next;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> close() {
+        if (!cancelled) {
+            synchronized (lock) {
+                if (!cancelled) {
+                    requestChainTail.completeExceptionally(new ClosedCursorException());
+
+                    cursorFut.thenAcceptAsync(cursor -> {
+                        if (cursor instanceof AutoCloseable) {
+                            try {
+                                ((AutoCloseable) cursor).close();
+
+                                cancelFut.complete(null);
+                            } catch (Exception e) {
+                                cancelFut.completeExceptionally(e);
+                            }
+                        } else {
+                            cancelFut.complete(null);
+                        }
+                    }, exec);
+
+                    cancelled = true;
+                }
+            }
+        }
+
+        return cancelFut.thenRun(() -> {});
```

Comment:
What is the purpose of no-op Runnable?

Reply:
To create an independent future which could be safely cancelled by the user.